File: leetcode/exercises/1.7.1.medium.product_of_array_except_self_2.js
// -----------------------
// Created     :7/11/2025
// Last Edited :7/11/2025 
// Topics      :
// Big O       :
// Problem Id  :38
// Source      :
// Notes       :
// -----------------------

/*
* @param {number[]} nums
* @return {number[]} 
*/
var productExceptSelf = function (nums) {

    res = Array.from({length: nums.length}, () => 1)

    prefix = 1
    for (let i = 0; i < nums.length; i++) { 
        res[i] = prefix
        prefix *= nums[i]
    }
    
    postfix = 1
    for (let i = nums.length - 1; i > 0; i--) {
    res[i-1] *= postfix*nums[i]
    postfix *= nums[i]
    }
    return res
}

console.log(productExceptSelf([1,2,3,4]))

File: leetcode/exercises/1.7.1.medium.product_of_array_except_self_2.go
// -----------------------
// Created     :7/11/2025
// Last Edited :7/11/2025 
// Topics      :
// Big O       :
// Problem Id  :38
// References  :
// Notes       :
// -----------------------

package main

import (
    "fmt"
)

func productExceptSelf(array []int) []int {

    res := []int{1}

    prefix := 1
    for i := 0; i < len(array)-1; i++ {
        res = append(res, prefix*array[i])
        prefix *= array[i]
    }

    postfix := 1
    for i := len(array)-1; i > -1; i-- {
        res[i] *= postfix
        postfix *= array[i]
    }

    return res

}

func main () {
    fmt.Println(productExceptSelf([]int{1,2,3,4}))
}
